craft_parts/sequencer.py

```diff
@@ -247,9 +290,56 @@ def _add_action(
             Action(part.name, step, action_type=action_type, reason=reason)
         )
 
+    def _ensure_overlay_consistency(
+        self, top_part: Part, reason: Optional[str] = None, skip_last: bool = False
+    ) -> LayerHash:
+        """Make sure overlay step layers are consistent.
+
+        The overlay step layers are stacked according to the part order. Each part
+        is given an identificaton value based on its overlay parameters and the value
+        of the previous layer in the stack, which is used to make sure the overlay
+        parameters for all previous layers remain the same. If any previous part
+        has not run, or had its parameters changed, it must run again to ensure
+        overlay consistency.
+
+        :param top_part: The part currently the top of the layer stack and whose
+            consistency is to be verified.
+        :param skip_last: Don't verify the consistency of the last (topmost) layer.
+            This is used during the overlay stack creation.
+
+        :return: This part's identification value.
+
+        :raise ValueError: If part is not in the project's parts list.
+        """
+        if top_part.name not in [p.name for p in self._part_list]:
+            raise ValueError(f"part {top_part!r} not in parts list")
+
+        for part in self._part_list:
+            layer_hash = self._layer_state.compute_layer_hash(part)
+
+            if skip_last and part.name == top_part.name:
+                return layer_hash
+
+            state_layer_hash = self._layer_state.get_layer_hash(part)
+
+            if layer_hash != state_layer_hash:
+                self._add_all_actions(
+                    target_step=Step.OVERLAY,
+                    part_names=[part.name],
+                    reason=reason,
+                )
+                self._layer_state.set_layer_hash(part, layer_hash)
+
+            if part.name == top_part.name:
+                return layer_hash
+
+        # execution should never reach this line
+        raise RuntimeError(f"part {top_part!r} not in parts list")
```

Comment:
I agree that code should never get here, but the reason is wrong (as it's checked that in line 314/315).

Then I started to dislike having two returns in the for loop.

So I tried to think how to reorder code so we don't have a "should not get here" situation, and came up with the following (untested, and just a suggestion):

```
        useful_part_list = []
        for part in self._part_list:
            useful_part_list.append(part)
            if top_part.name == part.name:
                if skip_last:
                    del useful_part_list[-1]
                break
        else:
            raise ValueError(f"part {top_part!r} not in parts list")

        # NOTE!! here is more evident what happens when there's an empty "useful part list",
        # which should be managed (it can happen when "top part is the first part and skip_last")
        # In the original code the situation is there, but more difficult to see, and it's 
        # returning a layer_hash that should NOT be returning.

        for part in useful_part_list:
            layer_hash = self._layer_state.compute_layer_hash(part)
            state_layer_hash = self._layer_state.get_layer_hash(part)

            if layer_hash != state_layer_hash:
                self._add_all_actions(
                    target_step=Step.OVERLAY,
                    part_names=[part.name],
                    reason=reason,
                )
                self._layer_state.set_layer_hash(part, layer_hash)

        return layer_hash

```

What do you think?

Reply:
We want to return the hash of the top layer, even if it's the first layer and verification against state is skipped (note that the skip_last parameter only skips the consistency check but the hash is still computed). I fixed the docstring to make this intent clearer, and slightly refactored the logic to remove one of the return points.

If we pre-build the useful part list and it ends up empty (because the top part is the first part and we skip last) we're not returning the top part's hash.

If I manage to further refactor the logic improve clarity I'll propose a follow-up PR.

Reply:
One more change: removed the initial test for part presence. It only fails in case of a programming error, and in this case we let the final guard to act (so it's not an "impossible" situation anymore).

Reply:
I see, thanks for all the clarifications.

One further detail, though: do we expect for all these hashes calculations and verifications to be super fast, or super slow?

Because if they will take a significative amount of time, maybe we should validate if the top_part is in the part list before doing everything else. Something similar to:

```
        useful_part_list = [] 
        for part in self._part_list: 
            useful_part_list.append(part) 
            if top_part.name == part.name: 
                break 
        else: 
            raise ValueError(f"part {top_part!r} not in parts list")

        for part in useful_part_list:
            layer_hash = self._layer_state.compute_layer_hash(part)

            # run the overlay step if the layer hash doesn't match the existing
            # state (unless we're in the top part and skipping the consistency check)
            if not (skip_last and part.name == top_part.name):
                state_layer_hash = self._layer_state.get_layer_hash(part)

                if layer_hash != state_layer_hash:
                    self._add_all_actions(
                        target_step=Step.OVERLAY,
                        part_names=[part.name],
                        reason=reason,
                    )
                    self._layer_state.set_layer_hash(part, layer_hash)
```

I just don't want to fall into those situations where the user makes a mistake, and the system takes a lot only to crash. It's an UX detail. But if it's fast, I'm happy with your code.